File: AS_JNI_Project/app/src/main/java/com/derry/as_jni_project/MainActivity.java
package com.derry.as_jni_project;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.widget.TextView;

// 生成头文件：javah com.derry.as_jni_project.MainActivity
public class MainActivity extends AppCompatActivity {
    static {
        System.loadLibrary("native-lib");
    }
    public static final int A = 100;
    public String name = "Derry"; // 签名：Ljava/lang/String;
    public static int age = 29; // 签名：I

    // Java 本地方法  实现：native层
    public native String getStringPwd();
    public static native String getStringPwd2();

    // -------------  交互操作 JNI
    public native void changeName();
    public static native void changeAge();
    public native void callAddMethod();

    // 专门写一个函数，给native成调用
    public int add(int number1, int number2) {
        return number1 + number2;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
         // Example of a call to a native method
        TextView tv = findViewById(R.id.sample_text);
        changeName();
        tv.setText(name);
        changeAge();
        tv.setText("" + age);
        callAddMethod();
    }

}

/*
    签名规则：
    Java的boolean  --- Z  注意点
    Java的byte  --- B
    Java的char  --- C
    Java的short  --- S
    Java的int  --- I
    Java的long  --- J     注意点
    Java的float  --- F
    Java的double  --- D
    Java的void  --- V
    Java的引用类型  --- Lxxx/xxx/xx/类名;
    Java的String  --- Ljava/lang/String;
    Java的array  int[]  --- [I
            double[][]  --- [[D
    int add(char c1, char c2) ---- (CC)I
    void a()     ===  ()V

    用命令查看签名：
    AS_JNI_Project/app/build/intermediates/javac/debug/classes/com/derry/as_jni_project
    javap -s -p xxx.class    -s 输出xxxx.class的所有属性和方法的签名，-p 忽略私有公开的所有属性方法全部输出
 */

File: ViewModuleDemo/jetpack_kotlin_project/src/main/java/com/example/jetpack_kotlin_project/MainActivity.kt
package com.example.jetpack_kotlin_project

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.ViewModelProvider
import com.example.jetpack_kotlin_project.databinding.ActivityMainBinding
import kotlin.concurrent.thread

/**
 * TODO 传统方式的
 *
 * 功能越多，代码越多，下面的缺点就越大
 *
 * 1.Activity 或 Fragment 是大管家，代码脓肿
 *
 * 2.Activity 要处理逻辑
 *
 * 3.Activity 要处理Model数据  UI 数据， 不仅管理了 却又管不好（横竖屏切换 数据丢失）
 *
 * 4.Activity 要实时刷新UI，例如：更新TextView
 *
 * 5.Activity 如果横竖屏切换 会丢失数据（号码数据一定会丢失的）
 *
 * 6.焊死，不能灵活拆卸
 *
 * ...... 等等
 *
 *
 *
 * TODO JetPack重构方式
 *  单元原则：只一件事情，管理绑定工作
 */

// 角色：MainActivity是管理者
class MainActivity : AppCompatActivity() {

    // 定义ViewModel
    private var vm: MainViewModel? = null

    // 定义我们的DataBinding
    private var dataBinding: ActivityMainBinding? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        dataBinding = DataBindingUtil.setContentView(this, R.layout.activity_main)

        // 新版本的写法，和旧版本不同
        vm = ViewModelProvider(viewModelStore, ViewModelProvider.AndroidViewModelFactory(application)).get(MainViewModel::class.java);

        // vm  和 DataBinding 绑定起来
        dataBinding!!.vm = vm

        //  DataBinding 感应  vm  <-----> 布局控件  计时改变
        dataBinding?.setLifecycleOwner(this)

        /*val name: String
        name.let {  }

        thread {

        }*/
    }
}
